### PR TITLE
fix: limit paginated queries by buffer size

### PR DESCRIPTION
**Pull request by pvlugter**
Refs #67

Query switching to and from backtracking relies on limited queries.

Will check this in tests against actual DynamoDB.

### CODE REVIEW

File: core/src/main/resources/reference.conf [hunk 1]
```diff
@@ -88,7 +88,10 @@ akka.persistence.dynamodb {
     }
 
     # In-memory buffer holding events when reading from DynamoDB.
-    buffer-size = 100
+    buffer-size = 1000
+
+    # Page size for paginated queries when reading from DynamoDB.
+    page-size = 100
```
Comment (pvlugter) [inline core/src/main/resources/reference.conf]: Add separate page size setting, and increase buffer size to 1000 (like r2dbc). Could also keep the buffer size at 100.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala [hunk 2]
```diff
@@ -198,32 +198,33 @@ import software.amazon.awssdk.services.dynamodb.model.QueryRequest
...
-                serId = item.get(EventSerId).n().toInt,
-                serManifest = "",
-                writerUuid = "", // not need in this query
-                tags = if (item.containsKey(Tags)) item.get(Tags).ss().asScala.toSet else Set.empty,
-                metadata = None)
-            } else {
-              createSerializedJournalItem(item, includePayload = true)
-            }
+        .mapConcat(_.items.iterator.asScala)
+        .take(settings.querySettings.bufferSize)
```
Comment (pvlugter) [inline core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:212]: Also checked that this doesn't end up pulling an extra page from the publisher. Looks as expected, with the mapConcat only pulling after the iterable is empty. So shouldn't be any extra queries if the page size and buffer size are the same, or multiples, and there haven't been any deleted events filtered out.

File: core/src/main/resources/reference.conf (hunk 1 above)
Comment (pvlugter) [reply core/src/main/resources/reference.conf]: May be better to keep the buffer size at 100. With potentially slower processing than r2dbc, queries could end up too far ahead of backtracking again.

File: core/src/main/scala/akka/persistence/dynamodb/DynamoDBSettings.scala [hunk 3]
```diff
@@ -81,6 +81,7 @@ final class QuerySettings(config: Config) {
   val backtrackingWindow: FiniteDuration = config.getDuration("backtracking.window").toScala
   val backtrackingBehindCurrentTime: FiniteDuration = config.getDuration("backtracking.behind-current-time").toScala
   val bufferSize: Int = config.getInt("buffer-size")
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/dynamodb/DynamoDBSettings.scala]: bufferSize is also used as the paging limit in eventsByPersistenceId. Should probably be changed to pageSize there as well so that we have a consistent use of page size config.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala [hunk 4]
```diff
@@ -198,32 +198,33 @@ import software.amazon.awssdk.services.dynamodb.model.QueryRequest
...
-                payload = None, // lazy loaded for backtracking
-                serId = item.get(EventSerId).n().toInt,
-                serManifest = "",
-                writerUuid = "", // not need in this query
-                tags = if (item.containsKey(Tags)) item.get(Tags).ss().asScala.toSet else Set.empty,
-                metadata = None)
-            } else {
-              createSerializedJournalItem(item, includePayload = true)
-            }
+        .mapConcat(_.items.iterator.asScala)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Wonder if we can go too far ahead of the backtracking window anyway. If the last event of the 1000 events (bufferSize) has a timestamp that is greater than previous backtracking timestamp + backtracking-window? In the query we have a toTimestamp, but for the normal query that is always based on current time. Maybe that toTimestamp should also be limited by where the previous backtracking query was.

I'm not sure this is needed since I think we always prioritize backtracking to catch up first.

File: core/src/main/resources/reference.conf (hunk 1 above)
Comment (franciscolopezsancho) [reply core/src/main/resources/reference.conf]: would it make sense as per [comment](https://github.com/lightbend/akka-persistence-dynamodb-private/pull/69#discussion_r1714670221) to add as a comment in this `reference.conf` that both, `buffer-size` and `page-size` should be the same or multiples. 
And maybe adding as well, as this [comment](https://github.com/lightbend/akka-persistence-dynamodb-private/pull/69#discussion_r1714752820) states that there are risks when setting the buffer size, too high?

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 4 above)
Comment (patriknw) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: However, we must also be aware of that there can be a period of time when there are no events at all, and that duration can be larger than the backtracking window.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 2 above)
Comment (patriknw) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:212]: Feels a little unnecessary to have two settings, but it's probably good. It's similar with jdbc/r2dbc queries. There is often a driver config for the result set paging, and then we set the sql limit.

File: core/src/main/scala/akka/persistence/dynamodb/internal/SnapshotDao.scala [hunk 5]
```diff
@@ -310,19 +310,20 @@ import software.amazon.awssdk.services.dynamodb.model.UpdateItemRequest
       .keyConditionExpression(s"$EntityTypeSlice = :entityTypeSlice AND $EventTimestamp BETWEEN :from AND :to")
       .expressionAttributeValues((attributeValues ++ filterAttributeValues).asJava)
       // Limit won't limit the number of results you get with the paginator.
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/dynamodb/internal/SnapshotDao.scala:312]: at least I noticed that the limit is about pagination 😄

**Review by patriknw — COMMENTED** [review]
good catch

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 2 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:212]: Yeah, probably unnecessary to have an extra setting. And Dynamo will limit pages by total size as well anyway.

File: core/src/main/scala/akka/persistence/dynamodb/DynamoDBSettings.scala (hunk 3 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/DynamoDBSettings.scala]: I think I already updated that too. But will probably go back to just buffer size anyway.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 4 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Yes, we could possibly get too far ahead with 1000 events, especially if the processing is slow enough.

Also thought about limiting the toTimestamp with the latest backtracking and window. But periods of time without events would need to be handled too, by moving forward in some way for empty results.

An alternative to limiting the end timestamp, would be doing it at the returned event level. So like pubsub skip if too far ahead, ignore events from regular query if they're too far ahead of the backtracking window, and process again once backtracking has caught up.

Comment (patriknw) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: That would mean that we would rely on that backtracking would catch the dropped events, and if there are many such dropped events that could make the performance situation even worse since backtracking events don't contain the payload, but will load the payload on demand when the event is accepted (not a duplicate).

Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Wouldn't be covered by backtracking, but by next regular query, as backtracking goes up to the latest query timestamp. Would just stop processing on the regular query if it's ahead by the backtracking window, then backtracking to catch up, then continue from where it was.

Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Would probably still need to recognise long pauses though, to avoid filtering.

Likely possible with r2dbc as well - under provisioned slow processing, with the query too far ahead on restart.

Comment (patriknw) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: I was thinking of dropping later in the stream, but I agree that it can be done inside the BySliceQuery

File: core/src/main/resources/reference.conf (hunk 1 above)
Comment (pvlugter) [reply core/src/main/resources/reference.conf]: Back to just buffer size setting at 100.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 2 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:212]: Removed the extra setting again. And buffer size back to 100.

**Comment by pvlugter** [comment]
Testing of this looks good.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 4 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: We probably do need this extra safeguard. I can look at it separate from the fix in this PR.

**Comment by pvlugter** [comment]
Buffer size back to 100 as before. Since these queries are per slice and then merged, the buffer size probably shouldn't be too big, as this is multiplied for each projection instance.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 4 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Trying to come up with a realistic scenario for this.

So to have a backtracking out-of-sequence error, we need to have had a delayed write, missed by the regular query, and for this to be further behind the latest query timestamp than the backtracking window (2 minutes) on restart.

Noting that it's only guaranteed to catch delayed events up to the backtracking behind-current-time, which is 10 seconds, when queries are keeping up-to-date. If events are delayed by more than 10 seconds, we could miss them regardless of restarts.

One way for the timestamp range in results to be more than the backtracking window is for the latest query timestamp to already be further behind, and events spread out enough. But if it's that far behind the write side (more than 2 minutes), we expect all delayed writes to already be there and will be seen by the regular query.

So to create this scenario, we need to have had queries be fairly up-to-date originally (within 10 seconds), miss a delayed write on the regular queries, and then keep processing slowly enough that it gets over 2 minutes ahead of that delayed write (and then restart before backtracking catches it). The toTimestamp is based on the current time before querying. Switch to backtracking is when more items than 3 x buffer size or more than half the backtracking window ahead. So it would need to have slow enough processing to stretch out more than 2 minutes, but without triggering backtracking first.

May be possible for DynamoDB — with the query streams being per slice, and then merged for the slice range, the downstream projection processing could be much slower relative to the buffer size and handling for each stream. We've also possibly seen something similar with a customer with r2dbc when under-provisioned.

Also thinking that there could be an optimisation for when queries are falling behind. Because we're only catching delayed events up to 10 seconds behind up-to-date backtracking, the expectation is that we won't have delayed writes further than this. If queries are even further behind the current time, then backtracking is not needed and could be disabled, reducing load to allow queries to better catch up, and which will see all events anyway.

Comment (patriknw) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Maybe the async update of the secondary index could be more delayed with DynamoDB? Some pids more delayed than others within the same slice. Maybe we should have more tolerance to with DynamoDB by increasing the default configuration for the 10 seconds and 2 minutes?

**Review by patriknw — APPROVED** [review]
LGTM

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 4 above)
Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Haven't seen any issues with secondary index delays. But should probably add some specific instrumentation for that, checking if we ever catch events with backtracking in large tests.

Tests so far with the fix in this PR all look good, with falling behind and projection restarts not causing issues. I'll move these ideas to a separate issue — to explore safeguarding against any edge cases we can think up and possible optimisations.

Comment (pvlugter) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala:211]: Moved to #70.
